File: static/js/solo/filters.js
// ========== filters.js ==========
// OneEuroFilter 类：用于平滑手部关键点，参数说明见构造函数
// KalmanFilter 类：1D 恒速卡尔曼滤波，对手部运动更友好

class KalmanFilter {
    /**
     * 1D 恒速卡尔曼滤波器 — 对运动追踪更友好
     * @param {number} t0 初始时间戳 (毫秒)
     * @param {number} x0 初始值
     * @param {number} q 过程噪声 (默认0.05，越大越信任测量值，越跟手)
     * @param {number} r 测量噪声 (默认0.5，越小越信任测量值)
     */
    constructor(t0, x0, q = 0.05, r = 0.5) {
        // 状态 [position, velocity]
        this.x = [x0, 0];
        // 协方差矩阵，初始有一定不确定性
        this.P = [[1, 0], [0, 1]];
        this.q = q;
        this.r = r;
        this.t_prev = t0;
    }

    filter(t, z) {
        const dt = (t - this.t_prev) / 1000; // 转为秒
        this.t_prev = t;
        if (dt <= 0.001) return this.x[0];

        // ---- 预测步骤 ----
        // F = [[1, dt], [0, 1]]
        const x_pred = [
            this.x[0] + this.x[1] * dt,
            this.x[1]
        ];
        // P_pred = F * P * F^T + Q
        const p00 = this.P[0][0] + dt * (this.P[1][0] + this.P[0][1]) + dt * dt * this.P[1][1];
        const p01 = this.P[0][1] + dt * this.P[1][1];
        const p10 = this.P[1][0] + dt * this.P[1][1];
        const p11 = this.P[1][1];
        // 加过程噪声 Q
        const dt2 = dt * dt;
        const dt3 = dt2 * dt;
        const dt4 = dt2 * dt2;
        const q = this.q;
        const P_pred = [
            [p00 + q * dt4 / 4, p01 + q * dt3 / 2],
            [p10 + q * dt3 / 2, p11 + q * dt2]
        ];

        // ---- 更新步骤 ----
        // H = [1, 0], 只观测位置
        const y = z - x_pred[0]; // 创新
        const S = P_pred[0][0] + this.r; // 创新协方差
        const K0 = P_pred[0][0] / S; // K[0]
        const K1 = P_pred[1][0] / S; // K[1]

        this.x[0] = x_pred[0] + K0 * y;
        this.x[1] = x_pred[1] + K1 * y;

        // P = (I - K*H) * P_pred
        const ikh00 = 1 - K0;
        this.P[0][0] = ikh00 * P_pred[0][0];
        this.P[0][1] = ikh00 * P_pred[0][1];
        this.P[1][0] = P_pred[1][0] - K1 * P_pred[0][0];
        this.P[1][1] = P_pred[1][1] - K1 * P_pred[0][1];

        return this.x[0];
    }
}

class OneEuroFilter {
    /**
     * @param {number} t0 初始时间戳 (毫秒)
     * @param {number} x0 初始值
     * @param {number} mincutoff 最小截止频率，控制平滑程度 (越小越平滑，默认 1.0)
     * @param {number} beta 速度系数，控制对快速运动的反应 (越大对快速运动越敏感，默认 0.03)
     * @param {number} dcutoff 导数截止频率 (默认 1.0，一般无需改动)
     */
    constructor(t0, x0, mincutoff = 1.0, beta = 0.03, dcutoff = 1.0) {
        this.mincutoff = mincutoff;
        this.beta = beta;
        this.dcutoff = dcutoff;
        this.x_prev = x0;
        this.dx_prev = 0.0;
        this.t_prev = t0;
    }

    smoothingFactor(t, cutoff) {
        const r = 2 * Math.PI * cutoff * t / 1000;
        return r / (r + 1);
    }

    exponentialSmoothing(a, x, x_prev) {
        return a * x + (1 - a) * x_prev;
    }

    filter(t, x) {
        const tElapsed = t - this.t_prev;
        if (tElapsed <= 0) return this.x_prev;
        const dx = (x - this.x_prev) / tElapsed;
        const a_d = this.smoothingFactor(tElapsed, this.dcutoff);
        const dx_hat = this.exponentialSmoothing(a_d, dx, this.dx_prev);
        const cutoff = this.mincutoff + this.beta * Math.abs(dx_hat);
        const a = this.smoothingFactor(tElapsed, cutoff);
        const x_hat = this.exponentialSmoothing(a, x, this.x_prev);
        this.x_prev = x_hat;
        this.dx_prev = dx_hat;
        this.t_prev = t;
        return x_hat;
    }
}

File: core/filters.py
import numpy as np


class KalmanFilter:
    """1D 恒速卡尔曼滤波器 — 对手部运动追踪更友好
    状态向量: [position, velocity]
    q: 过程噪声，越大越跟手 (建议 0.01~0.2)
    r: 测量噪声，越小越信任测量值 (建议 0.1~2.0)
    """
    def __init__(self, q=0.08, r=0.5):
        self.q = q
        self.r = r
        self.x = None  # [position, velocity]
        self.P = None  # 协方差矩阵
        self.last_time = None

    def __call__(self, z, timestamp=None):
        z = np.asarray(z, dtype=float)
        if timestamp is None:
            dt = 1.0 / 30
        else:
            if self.last_time is None:
                dt = 1.0 / 30
            else:
                dt = max(timestamp - self.last_time, 0.001)
        self.last_time = timestamp

        if self.x is None:
            # 首次初始化：位置=测量值，速度=0
            self.x = np.stack([z, np.zeros_like(z)], axis=-1)  # shape: (..., 2)
            self.P = np.tile(np.eye(2), z.shape + (1, 1))      # shape: (..., 2, 2)
            return z.copy()

        # ---- 预测 ----
        F = np.array([[1, dt], [0, 1]])
        x_pred = self.x @ F.T  # (..., 2)
        P_pred = F @ self.P @ F.T  # (..., 2, 2)
        # 加过程噪声 Q
        dt2, dt3, dt4 = dt * dt, dt * dt * dt, dt * dt * dt * dt
        Q = self.q * np.array([[dt4/4, dt3/2], [dt3/2, dt2]])
        P_pred = P_pred + Q

        # ---- 更新 ----
        H = np.array([1.0, 0.0])
        y = z - x_pred[..., 0]  # 创新
        S = P_pred[..., 0, 0] + self.r
        K0 = P_pred[..., 0, 0] / S
        K1 = P_pred[..., 1, 0] / S

        self.x[..., 0] = x_pred[..., 0] + K0 * y
        self.x[..., 1] = x_pred[..., 1] + K1 * y

        # 更新协方差
        ikh00 = 1 - K0
        self.P[..., 0, 0] = ikh00 * P_pred[..., 0, 0]
        self.P[..., 0, 1] = ikh00 * P_pred[..., 0, 1]
        self.P[..., 1, 0] = P_pred[..., 1, 0] - K1 * P_pred[..., 0, 0]
        self.P[..., 1, 1] = P_pred[..., 1, 1] - K1 * P_pred[..., 0, 1]

        return self.x[..., 0].copy()


class OneEuroFilter:
    """一欧元滤波器，用于平滑手部关键点轨迹，减少抖动"""
    def __init__(self, min_cutoff=0.5, beta=0.1, dcutoff=1.0):
        self.min_cutoff = min_cutoff
        self.beta = beta
        self.dcutoff = dcutoff
        self.x_prev = None
        self.dx_prev = None
        self.last_time = None

    def _alpha(self, cutoff, dt):
        tau = 1.0 / (2 * np.pi * cutoff)
        return 1.0 / (1.0 + tau / dt)

    def __call__(self, x, timestamp=None):
        x = np.asarray(x, dtype=float)
        if timestamp is None:
            dt = 1.0 / 30
        else:
            if self.last_time is None:
                dt = 1.0 / 30
            else:
                dt = timestamp - self.last_time
        self.last_time = timestamp

        if self.x_prev is None:
            self.x_prev = x.copy()
            self.dx_prev = np.zeros_like(x)
            return x.copy()

        dx = (x - self.x_prev) / dt
        a_d = self._alpha(self.dcutoff, dt)
        edx = a_d * dx + (1 - a_d) * self.dx_prev

        speed = np.linalg.norm(edx)
        cutoff = self.min_cutoff + self.beta * speed
        a = self._alpha(cutoff, dt)
        filtered = a * x + (1 - a) * self.x_prev

        self.x_prev = filtered
        self.dx_prev = edx
        return filtered